crates/bevy_pbr/src/light_probe/environment_filter.wgsl

```diff
@@ -0,0 +1,427 @@
+    let view = normal;
+    
+    // Get the roughness parameter
+    let roughness = constants.roughness;
+    
+    // Get blue noise offset for stratification
+    let blue_noise = blue_noise_offset(coords);
+    
+    var radiance = vec3f(0.0);
+    var total_weight = 0.0;
+    
+    // Skip sampling for mirror reflection (roughness = 0)
+    if (roughness < 0.01) {
+        radiance = sample_environment(normal, 0.0).rgb;
+        textureStore(output_texture, coords, face, vec4f(radiance, 1.0));
+        return;
+    }
+    
+    // For higher roughness values, use importance sampling
+    let sample_count = constants.sample_count;
+    
+    for (var i = 0u; i < sample_count; i++) {
+        // Get sample coordinates from Hammersley sequence with blue noise offset
+        var xi = hammersley_2d(i, sample_count);
+        xi = fract(xi + blue_noise); // Apply Cranley-Patterson rotation
+        
+        // Sample the GGX distribution to get a half vector
+        let half_vector = importance_sample_ggx(xi, roughness, normal);
+        
+        // Calculate reflection vector from half vector
+        let light_dir = reflect(-view, half_vector);
+        
+        // Calculate weight (N·L)
+        let NoL = dot(normal, light_dir);
+        
+        if (NoL > 0.0) {
+            // Calculate values needed for PDF
+            let NoH = dot(normal, half_vector);
+            let VoH = dot(view, half_vector);
+            let NoV = dot(normal, view);
+            
+            // Get the geometric shadowing term
+            let G = G_Smith(NoV, NoL, roughness);
+            
+            // Probability Distribution Function
+            let pdf = D_GGX(roughness, NoH) * NoH / (4.0 * VoH);
+            
+            // Calculate LOD using filtered importance sampling
+            // This is crucial to avoid fireflies and improve quality
+            let width = f32(size.x);
+            let lod = calculate_environment_map_lod(pdf, width, f32(sample_count));
+            
+            // Get source mip level - ensure we don't go negative
+            let source_mip = max(0.0, lod);
+            
+            // Sample environment map with the light direction
+            var sample_color = sample_environment(light_dir, source_mip).rgb;
+            sample_color = tonemap(sample_color);
+            
+            // Accumulate weighted sample, including geometric term
+            radiance += sample_color * NoL * G;
+            total_weight += NoL * G;
+        }
+    }
+    
+    // Normalize by total weight
+    if (total_weight > 0.0) {
+        radiance = radiance / total_weight;
+    }
+
+    // Reverse tonemap
+    radiance = reverse_tonemap(radiance);
+    
+    // Write result to output texture
+    textureStore(output_texture, coords, face, vec4f(radiance, 1.0));
+}
+
+@compute
+@workgroup_size(8, 8, 1)
+fn generate_irradiance_map(@builtin(global_invocation_id) global_id: vec3u) {
```

Comment:
FYI this uses the lambert BRDF, which is not what Bevy uses for StandardMaterial.

Generally I don't think it matters, we probably already assume lambert elsewhere, but you could technically importance sample the other BRDF if you did some more math.

Reply:
Thanks for flagging it. The diffuse (irradiance) cubemap is deliberately baked with simple cosine-weighted sampling: at run time StandardMaterial treats that value as plain irradiance and only direct lights get the full Burley term, while specular IBL relies on the split-sum trick (GGX’s view-independent D dot G baked into the radiance map, Fresnel applied per-pixel). 

Burley diffuse can’t be separated that way, since the N dot V factor ties it to the view vector. So baking it correctly would require an array of irradiance maps indexed by view angle/roughness and matching changes in every lighting path. Without that larger refactor, the Lambert-based bake is what Bevy expects.

ref https://github.com/bevyengine/bevy/blob/main/crates/bevy_pbr/src/light_probe/environment_map.wgsl#L256